// omnisharp-client/lib/omnisharp-client.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var candidate_finder_1 = require("./candidate-finder");
var client_v1_1 = require('./clients/client-v1');
var client_v2_1 = require('./clients/client-v2');
var composite_client_v1_1 = require('./clients/composite-client-v1');
var composite_client_v2_1 = require('./clients/composite-client-v2');
var enums = require("./enums");
exports.Driver = enums.Driver;
exports.DriverState = enums.DriverState;
exports.findCandidates = candidate_finder_1.findCandidates;
var OmnisharpClientV1 = (function (_super) {
    __extends(OmnisharpClientV1, _super);
    function OmnisharpClientV1() {
        _super.apply(this, arguments);
    }
    return OmnisharpClientV1;
})(client_v1_1.ClientV1);
exports.OmnisharpClientV1 = OmnisharpClientV1;
var OmnisharpClientV2 = (function (_super) {
    __extends(OmnisharpClientV2, _super);
    function OmnisharpClientV2() {
        _super.apply(this, arguments);
    }
    return OmnisharpClientV2;
})(client_v2_1.ClientV2);
exports.OmnisharpClientV2 = OmnisharpClientV2;
var OmnisharpObservationClientV1 = (function (_super) {
    __extends(OmnisharpObservationClientV1, _super);
    function OmnisharpObservationClientV1() {
        _super.apply(this, arguments);
    }
    return OmnisharpObservationClientV1;
})(composite_client_v1_1.ObservationClientV1);
exports.OmnisharpObservationClientV1 = OmnisharpObservationClientV1;
var OmnisharpCombinationClientV1 = (function (_super) {
    __extends(OmnisharpCombinationClientV1, _super);
    function OmnisharpCombinationClientV1() {
        _super.apply(this, arguments);
    }
    return OmnisharpCombinationClientV1;
})(composite_client_v1_1.CombinationClientV1);
exports.OmnisharpCombinationClientV1 = OmnisharpCombinationClientV1;
var OmnisharpObservationClientV2 = (function (_super) {
    __extends(OmnisharpObservationClientV2, _super);
    function OmnisharpObservationClientV2() {
        _super.apply(this, arguments);
    }
    return OmnisharpObservationClientV2;
})(composite_client_v2_1.ObservationClientV2);
exports.OmnisharpObservationClientV2 = OmnisharpObservationClientV2;
var OmnisharpCombinationClientV2 = (function (_super) {
    __extends(OmnisharpCombinationClientV2, _super);
    function OmnisharpCombinationClientV2() {
        _super.apply(this, arguments);
    }
    return OmnisharpCombinationClientV2;
})(composite_client_v2_1.CombinationClientV2);
exports.OmnisharpCombinationClientV2 = OmnisharpCombinationClientV2;
