// omnisharp-client/lib/clients/composite-client-base.js
var rx_1 = require("rx");
var _ = require('lodash');
var ObservationClientBase = (function () {
    function ObservationClientBase(clients) {
        var _this = this;
        if (clients === void 0) { clients = []; }
        this.clients = clients;
        this._disposable = new rx_1.CompositeDisposable();
        this._clientDisposable = new rx_1.CompositeDisposable();
        this._clientsSubject = new rx_1.ReplaySubject(1);
        this.onNext = function () { return _this._clientsSubject.onNext(_this.clients.slice()); };
        this.projectAdded = this.makeMergeObserable(function (client) { return client.projectAdded; });
        this.projectChanged = this.makeMergeObserable(function (client) { return client.projectChanged; });
        this.projectRemoved = this.makeMergeObserable(function (client) { return client.projectRemoved; });
        this.error = this.makeMergeObserable(function (client) { return client.error; });
        this.msBuildProjectDiagnostics = this.makeMergeObserable(function (client) { return client.msBuildProjectDiagnostics; });
        this.packageRestoreStarted = this.makeMergeObserable(function (client) { return client.packageRestoreStarted; });
        this.packageRestoreFinished = this.makeMergeObserable(function (client) { return client.packageRestoreFinished; });
        this.unresolvedDependencies = this.makeMergeObserable(function (client) { return client.unresolvedDependencies; });
        this.events = this.makeMergeObserable(function (client) { return client.events; });
        this.commands = this.makeMergeObserable(function (client) { return client.commands; });
        this.state = this.makeMergeObserable(function (client) { return client.state; });
        this.status = this.makeMergeObserable(function (client) { return client.status; });
        this.requests = this.makeMergeObserable(function (client) { return client.requests; });
        this.responses = this.makeMergeObserable(function (client) { return client.responses; });
        this.errors = this.makeMergeObserable(function (client) { return client.errors; });
        this.onNext();
        this._disposable.add(this._clientsSubject);
        this._disposable.add(this._clientDisposable);
    }
    ObservationClientBase.prototype.dispose = function () {
        if (this._disposable.isDisposed)
            return;
        this._disposable.dispose();
    };
    ObservationClientBase.prototype.makeMergeObserable = function (selector) {
        return this._clientsSubject.flatMapLatest(function (clients) { return rx_1.Observable.merge.apply(rx_1.Observable, clients.map(selector)); }).share();
    };
    ObservationClientBase.prototype.observe = function (selector) {
        return this.makeMergeObserable(selector);
    };
    ObservationClientBase.prototype.add = function (client) {
        var _this = this;
        this.clients.push(client);
        this.onNext();
        var d = rx_1.Disposable.create(function () {
            _.pull(_this.clients, client);
            _this.onNext();
        });
        this._clientDisposable.add(d);
        return d;
    };
    return ObservationClientBase;
})();
exports.ObservationClientBase = ObservationClientBase;
var CombinationClientBase = (function () {
    function CombinationClientBase(clients) {
        var _this = this;
        if (clients === void 0) { clients = []; }
        this.clients = clients;
        this._disposable = new rx_1.CompositeDisposable();
        this._clientDisposable = new rx_1.CompositeDisposable();
        this._clientsSubject = new rx_1.ReplaySubject(1);
        this.onNext = function () { return _this._clientsSubject.onNext(_this.clients.slice()); };
        this.projectAdded = this.makeCombineObserable(function (client) { return client.projectAdded; });
        this.projectChanged = this.makeCombineObserable(function (client) { return client.projectChanged; });
        this.projectRemoved = this.makeCombineObserable(function (client) { return client.projectRemoved; });
        this.error = this.makeCombineObserable(function (client) { return client.error; });
        this.msBuildProjectDiagnostics = this.makeCombineObserable(function (client) { return client.msBuildProjectDiagnostics; });
        this.packageRestoreStarted = this.makeCombineObserable(function (client) { return client.packageRestoreStarted; });
        this.packageRestoreFinished = this.makeCombineObserable(function (client) { return client.packageRestoreFinished; });
        this.unresolvedDependencies = this.makeCombineObserable(function (client) { return client.unresolvedDependencies; });
        this.state = this.makeCombineObserable(function (client) { return client.state; });
        this.status = this.makeCombineObserable(function (client) { return client.status; });
        this.onNext();
        this._disposable.add(this._clientsSubject);
        this._disposable.add(this._clientDisposable);
    }
    CombinationClientBase.prototype.dispose = function () {
        if (this._disposable.isDisposed)
            return;
        this._disposable.dispose();
    };
    CombinationClientBase.prototype.makeCombineObserable = function (selector) {
        var cache = {};
        return this._clientsSubject.flatMapLatest(function (clients) {
            var removal = _.difference(_.keys(cache), clients.map(function (z) { return z.uniqueId; }));
            _.each(removal, function (z) { return delete cache[z]; });
            return rx_1.Observable.combineLatest(clients.map(function (z) { return selector(z).startWith(cache[z.uniqueId]); }), function () {
                var values = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    values[_i - 0] = arguments[_i];
                }
                return values.map(function (value, index) {
                    cache[clients[index].uniqueId] = value;
                    return {
                        key: clients[index].uniqueId,
                        value: value
                    };
                });
            });
        }).share();
    };
    CombinationClientBase.prototype.observe = function (selector) {
        return this.makeCombineObserable(selector);
    };
    CombinationClientBase.prototype.add = function (client) {
        var _this = this;
        this.clients.push(client);
        this.onNext();
        var d = rx_1.Disposable.create(function () {
            _.pull(_this.clients, client);
            _this.onNext();
        });
        this._clientDisposable.add(d);
        return d;
    };
    return CombinationClientBase;
})();
exports.CombinationClientBase = CombinationClientBase;
